spec/controllers/projects/classifications_controller_spec.rb

```diff
@@ -0,0 +1,208 @@
+    end
+  end
+
+  shared_context 'request can be found' do
+    include_context 'project can be found'
+
+    let(:info_request) { FactoryBot.create(:info_request, user: user) }
+
+    before do
+      info_requests = double(:info_requests_collection)
+      allow(project).to receive(:info_requests).and_return(info_requests)
+      allow(info_requests).to receive(:find_by!).with(url_title: 'url_title').
+        and_return(info_request)
+    end
+  end
+
+  describe '#create' do
+    let(:user) { FactoryBot.create(:pro_user) }
+    let(:ability) { Object.new.extend(CanCan::Ability) }
+
+    before do
+      session[:user_id] = user.id
+      allow(controller).to receive(:current_user).and_return(user)
+
+      allow(controller).to receive(:current_ability).and_return(ability)
+    end
+
+    context 'project to be classified can not be found' do
+      it'raises a ActiveRecord::RecordNotFound error' do
+        expect {
+          post :create, params: { project_id: 'invalid' }
+        }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+
+    context 'request to be classified can not be found' do
+      include_context 'project can be found'
+
+      it'raises a ActiveRecord::RecordNotFound error' do
+        expect {
+          post :create, params: { project_id: '1', url_title: 'invalid' }
+        }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+
+    shared_context 'request to be classified can be found' do
+      include_context 'request can be found'
+
+      def post_status(status, message: nil)
+        classification = { described_state: status }
+        classification[:message] = message if message
+
+        post :create, params: {
+          classification: classification,
+          project_id: '1',
+          url_title: 'url_title'
+        }
+      end
+    end
+
+    context 'user is not allow to update the request' do
+      include_context 'request to be classified can be found'
+
+      before { ability.cannot :update_request_state, info_request }
+
+      it 'raises a CanCan::AccessDenied error' do
+        expect {
+          post_status('successful')
+        }.to raise_error(CanCan::AccessDenied)
+      end
+    end
+
+    shared_context 'user can classify request' do
+      include_context 'request to be classified can be found'
+      before { ability.can :update_request_state, info_request }
+    end
+
+    context 'user is allowed to update the request' do
+      include_context 'user can classify request'
+
+      it 'should create status_update log' do
+        post_status('successful')
+
+        event = assigns(:status_update_event)
+        expect(event).to be_a InfoRequestEvent
+        expect(event.event_type).to eq 'status_update'
+        expect(event.params[:described_state]).to eq 'successful'
+        expect(event.params[:old_described_state]).to eq 'waiting_response'
+        expect(event.params[:user_id]).to eq info_request.user_id
```

Comment:
Ah we'll want to record the _actual_ user who classified the request, rather than assuming its the request owner.

Perhaps we want to break these up more like:

```ruby
context 'a project owner classifies their own request'
  # …

context "a project contributor classifies an owner's request"
  # …
```

Our implementation does use [`current_user.id`](https://github.com/mysociety/alaveteli/blob/4b02ca7ee892688bcea18cd9a32579d6e3fb5cc1/app/controllers/concerns/classifiable.rb#L78), but this spec doesn't illustrate that intent.

Might be worth reviewing what we do for the classification game here and whether we want to create a [`RequestClassification`](https://github.com/mysociety/alaveteli/blob/4b02ca7ee892688bcea18cd9a32579d6e3fb5cc1/app/controllers/classifications_controller.rb#L44-L64).